### PR TITLE
interfaces/builtin: updates to dcdbas-control interface

### PR DESCRIPTION
**Pull request by tonyespy**
Added additional /dev and /proc nodes to the interface.
Also added the iopl (set IO privilege) system call to
the interface's ConnectedPlugSecComp.

### CODE REVIEW

File: interfaces/builtin/dcdbas_control.go [hunk 1]
```diff
@@ -45,13 +45,30 @@ const dcdbasControlConnectedPlugAppArmor = `
 /sys/devices/platform/dcdbas/host_control_action rw,
 /sys/devices/platform/dcdbas/host_control_smi_type rw,
 /sys/devices/platform/dcdbas/host_control_on_shutdown rw,
+
+# additional devices required by dchbas library
+/dev/EsmBASDev r,
+/dev/port r,
+/dev/mem r,
```
Comment (morphis) [inline interfaces/builtin/dcdbas_control.go]: This makes this super privileged and makes me thinking if we shouldn't have a separate interface like 'memory-observe' for such memory access. @jdstrand what do you think about this?

Furthermore dcdbas is specific to Dell devices but an official kernel interface as described in https://www.kernel.org/doc/Documentation/dcdbas.txt so I am not sure if its correct to merge library implementation specifics in here.

**Comment by tonyespy** [comment]
dcdbas_control was added specifically for Dell's DCM snap to allow it to handle confined communication with their BIOS implementation.  I've asked for detailed use cases for each of the extra devices that were added in this pull-request, but my query has yet to be answered.  I would have preferred to be able to have a single interface only usable by Dell for this DCM snap, however as this approach has been NAK'd, I'll close this one out, and work on separate interface changes for each of the additional device nodes and the iopl system call.

@jdstrand will you handle the generic physical-mem-observable interface?

**Comment by jdstrand** [comment]
/dev/port is dangerous. From http://www.tldp.org/HOWTO/IO-Port-Programming-2.html: "Naturally, for this to work your program needs read/write access to /dev/port. This method is probably slower than the normal method above, but does not need compiler optimisation nor ioperm(). It doesn't need root access either, if you give a non-root user or group access to /dev/port --- but this is a very bad thing to do in terms of system security, since it is possible to hurt the system, perhaps even gain root access, by using /dev/port to access hard disks, network cards, etc. directly." Before /dev/port is added, we need to know a bit more about how it is being used.

/dev/mem is indeed separate. We've seen a few cases of things needing it and we discussed this via email. I'll paste the relevant part of that email from me here:

```
The configuration item CONFIG_STRICT_DEVMEM[1]:

"If this option is disabled, you allow userspace (root) access to all
of memory, including kernel and userspace memory. Accidental
access to this is obviously disastrous, but specific access can
be used by people debugging the kernel. Note that with PAT support
enabled, even in this case there are restrictions on /dev/mem
use due to the cache aliasing requirements.

If this option is switched on, the /dev/mem file only allows
userspace access to PCI space and the BIOS code and data regions.
This is sufficient for dosemu and X and all common users of
/dev/mem."

We probably need to ensure that CONFIG_STRICT_DEVMEM=y for snappy kernel snaps
as part of the porting guide[2]. It looks like it is present in
security.config[3]. Tyler, is this all that is needed to ensure
CONFIG_STRICT_DEVMEM=y in snappy kernels?

Still, even when CONFIG_STRICT_DEVMEM=y, it allows a lot of privileged access
and my understanding on non-x86 architectures the restrictions are less
clear[4]. Access to /dev/mem has come up in other places, but we've typically
said 'no'. However, considering CONFIG_STRICT_DEVMEM=y, I think that this is a
candidate for an interface (fyi, '/dev/mem r,' is already in the kernel-module-
control interface since lscpu needs it). Well need it for an X or dosemu snap at
a minimum....

Interestingly, kernel-module-control doesn't allow CAP_SYS_RAWIO and man
capabilities indicates it is needed for opening /dev/mem. I suspect that
CAP_SYS_RAWIO is only needed for writing to /dev/mem. 

...

Tyler, I suspect that we are going to need to have two interfaces:

 * physical-memory-observe (uses 'deny-auto-connection: true' in base 
   declaration)

     # With kernels with STRICT_DEVMEM=n, read-only access to all physical
     # memory. With STRICT_DEVMEM=y, allow reading /dev/mem for read-only 
     # access to architecture-specific subset of the physical address (eg, PCI, 
     # space, BIOS code and data regions on x86, etc)
     /dev/mem r,

 * physical-memory-control (use 'deny-auto-connection: true' and 'deny-
   installation: true' in base declaration)

     # With kernels with STRICT_DEVMEM=n, write access to all physical
     # memory.
     #
     # With STRICT_DEVMEM=y, allow writing to /dev/mem to access
     # architecture-specific subset of the physical address (eg, PCI space,
     # BIOS code and data regions on x86, etc) for all common uses of /dev/mem
     # (eg, X without KMS, dosemu, etc)
     capability sys_rawio,
     /dev/mem rw,

[1]https://www.kernel.org/doc/menuconfig/arch-x86-Kconfig.debug.html#STRICT_DEVMEM
[2]https://docs.google.com/document/d/1LOqrSty2yTsUtQ4nFehaisZc4XmKKsK_E_tz-O1hU-M/edit
[3]https://git.launchpad.net/~p-pisati/ubuntu/+source/linux/tree/kernel/configs/snappy/security.config?h=snappy_v4.4
[4]http://lists.infradead.org/pipermail/linux-arm-kernel/2014-November/299833.html
```

Tony, I think I gave everything needed for what to put in the physical-memory-observe and physical-memory-control interfaces. Feel free to create both in the same PR. If you need me to do it, please talk to @tyhicks on priorities.

Lastly, I'm unfamiliar with /dev/EsmBASDev but it seems like handling it could be done by this interface based on: https://gitweb.gentoo.org/dev/liquidx.git/diff/app-admin/omsa/files/4.5/dellomsa-drivers?id=be1511b955b0441f0ae007896bc65d718c35b563

However it isn't at all clear why it needs access to /dev/EsmBASDev. Should this be handled by the udev backend, the kmod backend, the apparmor backend? Can you provide more information on why this access is needed and then we can advise on how to make it available.

Thanks!

**Comment by tonyespy** [comment]
@jdstrand 

Thanks for the response.  I went back through each of the devices:
 
* /dev/EsmBASDev - belongs to the dcdesm driver.  Neither the device nor the kernel module are available on stlouis UC16 devices, so I've asked whether or not this is really a requirement.  I actually stumbled across the same page you had.  Looks like there's also dcdtvm and dcdipm modules.  If this device is required, then we should probably add to a dedicated dcdesm_control or _observe interface.

 * /dev/port - I've asked again for details on this.  Note, /dev/port is covered by the same man page as /dev/mem [1], so perhaps it belongs in the /dev/mem interfaces.  Regarding iopl, the same page you pointed out only mentions it being required when using the  port IO macros defined in /usr/include/asm/io.h, however maybe it's also required for access via /dev/port?

 * /dev/mem - I'll get going on a pull-request.

 * /dev/char/mem/* - none of these devices exist on my stlouis.  I've asked for more clarification.

 * /proc/bus/pci - I've suggested use of the hardware_observe interface.  If they need this for something other then access to /proc/bus/pci/devices, we'll address when we have more information.

[1] http://man7.org/linux/man-pages/man4/mem.4.html

**Comment by jdstrand** [comment]
Re /dev/port> that's interesting, but I think it is sufficiently different from 'physical-memory-control' to warrant its own interface. Can you create `io-ports-control` with the following rule: `/dev/port rw,`. I don't know if io-ports-observe is interesting like physical-memory-observe is for lscpu so let's skip that one (unless you or someone else has a use case for it). Feel free to add physical-memory-control, physical-memory-observe and io-ports-control as part of the same PR if that makes sense for you.

**Pull request closed without merging**